pkg/config/config.go

```diff
@@ -1042,6 +1042,15 @@ func findUnknownKeys(config Config) []string {
 func load(config Config, origin string, loadSecret bool) (*Warnings, error) {
 	warnings := Warnings{}
 
+	// Feature detection running in a defer func as it always  need to run (whether config load has been successful or not)
+	// Because some Agents (e.g. trace-agent) will run even if config file does not exist
+	defer func() {
+		// Environment feature detection needs to run before applying override funcs
+		// as it may provide such overrides
+		DetectFeatures()
+		applyOverrideFuncs(config)
+	}()
+
```

Comment:
This does not necessarily guarantee that feature detection will run by any particular point in time.

I'm not too familiar with the history here, so I'm guessing that you haven't just moved these two calls to this point in the function (without spawning a goroutine) because they take a while to run.

If so, perhaps a `sync.WaitGroup` or similar could be used so that you can start the detection goroutine here (early) and then wait for it to complete before it's needed (at least below around line 1077, where the `DetectFeatures` call used to be, and probably somewhere in trace-agent as well.

If not, then I'm totally off base :)

Reply:
Go guarantees the `defer` function is run after function returns, so it will always run while before it would run only if the config file is read successfully.

Reply:
Sorry, I somehow read the `defer` there as `go`.  So yeah, my comment is totally off-base.